File: backend/internal/vectorstore/vectorstore.go
package vectorstore

import (
	_ "embed"
	"encoding/json"
	"fmt"
	"math"
	"sort"
)

//go:embed data/images.json
var imagesData []byte

//go:embed data/directions.json
var directionsData []byte

//go:embed data/demo.json
var demoData []byte

type ImageEntry struct {
	URL       string    `json:"url"`
	Embedding []float32 `json:"embedding"`
}

type DirectionVectors struct {
	Energy  []float32 `json:"energy"`
	Valence []float32 `json:"valence"`
	Tempo   []float32 `json:"tempo"`
	Texture []float32 `json:"texture"`
}

type DemoData struct {
	Embedding   []float32     `json:"embedding"`
	MoodEnergy  float32       `json:"mood_energy"`
	MoodValence float32       `json:"mood_valence"`
	MoodTempo   float32       `json:"mood_tempo"`
	MoodTexture float32       `json:"mood_texture"`
	Images      []ImageResult `json:"images"`
}

type ImageResult struct {
	ID       int     `json:"id"`
	ImageURL string  `json:"image_url"`
	Score    float32 `json:"score"`
}

type VectorStore struct {
	images     []ImageEntry
	directions DirectionVectors
	demo       DemoData
}

func New() (*VectorStore, error) {
	vs := &VectorStore{}

	if err := json.Unmarshal(imagesData, &vs.images); err != nil {
		return nil, fmt.Errorf("failed to parse images.json: %w", err)
	}

	if err := json.Unmarshal(directionsData, &vs.directions); err != nil {
		return nil, fmt.Errorf("failed to parse directions.json: %w", err)
	}

	if err := json.Unmarshal(demoData, &vs.demo); err != nil {
		return nil, fmt.Errorf("failed to parse demo.json: %w", err)
	}

	return vs, nil
}

func (vs *VectorStore) Search(embedding []float32, topK int) []ImageResult {
	type scored struct {
		index int
		dist  float32
	}

	results := make([]scored, 0, len(vs.images))
	for i, img := range vs.images {
		dist := l2Distance(embedding, img.Embedding)
		results = append(results, scored{index: i, dist: dist})
	}

	sort.Slice(results, func(i, j int) bool {
		return results[i].dist < results[j].dist
	})

	if topK > len(results) {
		topK = len(results)
	}

	out := make([]ImageResult, topK)
	for i := 0; i < topK; i++ {
		idx := results[i].index
		out[i] = ImageResult{
			ID:       idx,
			ImageURL: vs.images[idx].URL,
			Score:    results[i].dist,
		}
	}

	return out
}

// RefineEmbedding applies mood slider adjustments to a base embedding.
// This is an exact port of bridge.py:87-113.
func (vs *VectorStore) RefineEmbedding(base []float32, energy, valence, tempo, texture float32) []float32 {
	dim := len(base)
	refined := make([]float32, dim)
	copy(refined, base)

	for i := 0; i < dim; i++ {
		var adjustment float32
		if i < len(vs.directions.Energy) {
			adjustment += vs.directions.Energy[i] * (energy - 0.5) * 0.2
		}
		if i < len(vs.directions.Valence) {
			adjustment += vs.directions.Valence[i] * (valence - 0.5) * 0.2
		}
		if i < len(vs.directions.Tempo) {
			adjustment += vs.directions.Tempo[i] * (tempo - 0.5) * 0.15
		}
		if i < len(vs.directions.Texture) {
			adjustment += vs.directions.Texture[i] * (texture - 0.5) * 0.15
		}
		refined[i] += adjustment
	}

	// L2 normalize
	var norm float64
	for _, v := range refined {
		norm += float64(v) * float64(v)
	}
	norm = math.Sqrt(norm)
	if norm > 0 {
		for i := range refined {
			refined[i] = float32(float64(refined[i]) / norm)
		}
	}

	return refined
}

func (vs *VectorStore) GetDemo() *DemoData {
	return &vs.demo
}

func (vs *VectorStore) ImageCount() int {
	return len(vs.images)
}

func l2Distance(a, b []float32) float32 {
	var sum float64
	n := len(a)
	if len(b) < n {
		n = len(b)
	}
	for i := 0; i < n; i++ {
		d := float64(a[i]) - float64(b[i])
		sum += d * d
	}
	return float32(math.Sqrt(sum))
}

File: worker/src/lib/vectorstore.ts
import type { ImageEntry, DirectionVectors, DemoData, ImageResult } from './types';
import imagesData from '../data/images.json';
import directionsData from '../data/directions.json';
import demoData from '../data/demo.json';

const images: ImageEntry[] = imagesData as ImageEntry[];
const directions: DirectionVectors = directionsData as DirectionVectors;
const demo: DemoData = demoData as DemoData;

function l2Distance(a: number[], b: number[]): number {
  let sum = 0;
  const n = Math.min(a.length, b.length);
  for (let i = 0; i < n; i++) {
    const d = a[i] - b[i];
    sum += d * d;
  }
  return Math.sqrt(sum);
}

export function search(embedding: number[], topK: number): ImageResult[] {
  const scored: { index: number; dist: number }[] = [];
  for (let i = 0; i < images.length; i++) {
    const dist = l2Distance(embedding, images[i].embedding);
    scored.push({ index: i, dist });
  }

  scored.sort((a, b) => a.dist - b.dist);

  const k = Math.min(topK, scored.length);
  const results: ImageResult[] = [];
  for (let i = 0; i < k; i++) {
    const idx = scored[i].index;
    results.push({
      id: idx,
      image_url: images[idx].url,
      score: scored[i].dist,
    });
  }

  return results;
}

export function refineEmbedding(
  base: number[],
  energy: number,
  valence: number,
  tempo: number,
  texture: number,
): number[] {
  const dim = base.length;
  const refined = base.slice();

  for (let i = 0; i < dim; i++) {
    let adjustment = 0;
    if (i < directions.energy.length) {
      adjustment += directions.energy[i] * (energy - 0.5) * 0.2;
    }
    if (i < directions.valence.length) {
      adjustment += directions.valence[i] * (valence - 0.5) * 0.2;
    }
    if (i < directions.tempo.length) {
      adjustment += directions.tempo[i] * (tempo - 0.5) * 0.15;
    }
    if (i < directions.texture.length) {
      adjustment += directions.texture[i] * (texture - 0.5) * 0.15;
    }
    refined[i] += adjustment;
  }

  // L2 normalize
  let norm = 0;
  for (const v of refined) {
    norm += v * v;
  }
  norm = Math.sqrt(norm);
  if (norm > 0) {
    for (let i = 0; i < refined.length; i++) {
      refined[i] /= norm;
    }
  }

  return refined;
}

export function getDemo(): DemoData {
  return demo;
}

export function imageCount(): number {
  return images.length;
}
